akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/CommandHandler.scala

```diff
@@ -23,23 +25,130 @@ trait CommandHandler[Command, Event, State] {
 }
 
 object CommandHandlerBuilder {
+  def builder[Command, Event, State](): CommandHandlerBuilder[Command, Event, State] =
+    new CommandHandlerBuilder[Command, Event, State]
+}
+
+final class CommandHandlerBuilder[Command, Event, State]() {
+
+  private var builders: List[CommandHandlerBuilderByState[Command, Event, State, State]] = Nil
 
   /**
+   * Use this method to define command handlers that are selected when the passed predicate holds true.
+   *
+   * @param statePredicate The handlers defined by this builder are used when the `statePredicate` is `true`
+   *
+   * @return A new, mutable, CommandHandlerBuilderByState
+   */
+  def forState(statePredicate: Predicate[State]): CommandHandlerBuilderByState[Command, Event, State, State] = {
+    val builder = CommandHandlerBuilderByState.builder[Command, Event, State](statePredicate)
+    builders = builder :: builders
+    builder
+  }
+
+  /**
+   * Use this method to define command handlers that are selected when the passed predicate holds true
+   * for a given subtype of your model. Useful when the model is defined as class hierarchy.
+   *
    * @param stateClass The handlers defined by this builder are used when the state is an instance of the `stateClass`
-   * @return A new, mutable, command handler builder
+   * @param statePredicate The handlers defined by this builder are used when the `statePredicate` is `true`
+   *
+   * @return A new, mutable, CommandHandlerBuilderByState
    */
-  def builder[Command, Event, S <: State, State](stateClass: Class[S]): CommandHandlerBuilder[Command, Event, S, State] =
-    new CommandHandlerBuilder(statePredicate = new Predicate[S] {
-      override def test(state: S): Boolean = state != null && stateClass.isAssignableFrom(state.getClass)
-    })
+  def forState[S <: State](stateClass: Class[S], statePredicate: Predicate[S]): CommandHandlerBuilderByState[Command, Event, S, State] = {
```

Comment:
Was unsure about that old scenario with type bounds and Java so I double checked that this bound is enforced from Java as well. 👍